compiler/codegen/expr.cpp

```diff
@@ -2497,41 +2387,170 @@ GenRet codegenCallExpr(GenRet function,
       const clang::CodeGen::ABIArgInfo& returnInfo = CGI->getReturnInfo();
       returnInfo.canHaveCoerceToType();
 
-      // We might be doing 'structure return'
-      // TODO: use CGI info
-      if( fnType->getReturnType()->isVoidTy() &&
-          fnType->getNumParams() >= 1 &&
-          func && func->hasStructRetAttr() ) {
-        // We must allocate a temporary to store the return value
-        llvm::PointerType* ptrToRetTy = llvm::cast<llvm::PointerType>(
-            fnType->getParamType(0));
-        llvm::Type* retTy = ptrToRetTy->getElementType();
-        sret = createVarLLVM(retTy);
+      if (CGI->getArgStruct() != NULL)
+        INT_FATAL("inalloca arguments not yet implemented");
+
+      if (returnInfo.isIndirect()) {
+        // Create a temporary for holding the return value
+        llvm::Type* chapelRetTy = fn->retType->codegen().type;
+        sret = createVarLLVM(chapelRetTy);
         llArgs.push_back(sret);
       }
-
     }
 
+    ArgSymbol* formal = NULL;
+    if (fn && fn->numFormals() > 0)
+      formal = fn->getFormal(1);
```

Comment:
Can't we track formals inside the loop below?

Reply:
I'm not sure what you mean? The loop below does move on to the next formal. But it can't use `for_formals` e.g. because it's possible that `fn` is NULL.

Reply:
It took me a search to find that you are moving to the next formal, and I was kind of thinking that there is something special with the first formal (i.e. is it `this`? are we looking at a method?)

So, I was just thinking that within the for loop below, you could do

```
if (fn) && fn->numFormals() > 0) {
  formal = fn->getFormal(i+1);
}
```
But I am not sure about its implications.

It is me nitpicking and the point is mostly moot now, anyways.

Reply:
Right, I didn't do that because I didn't want to do an O(n) search in an O(n) loop (resulting in O(n*n)). (But maybe the solution to that is to make getFormal O(1)...)

Reply:
I feel like I am having a dejavu... I think you have told me about getFormal being O(n) before but I forgot. It is unfortunate but this implementation makes more sense, then. Thanks!